### DISCUSSION TITLE
Hydration errors after updating to latest Next js version

### DISCUSSION THREAD

**Discussion by vishalshah6153** [vercel/next.js#42728] [Help]
Recently, I've updated the nextjs version from 12 to 13. And it turns out that there are a lot of Hydration errors. I can go through them and try to resolve them but there are some cases in that I need the HTML structure to be the way it is. And going through each file and resolving the errors doesn't seem to be an efficient way. So is there any way to avoid these errors or resolve them all at once?

![image](https://user-images.githubusercontent.com/89897183/201025696-6e3e646d-c791-49be-b6bd-ed11c919cb54.png)

**Comment by icyJoseph** [comment]
Hi,

This is pretty much because the first HTML frame rendered on the client has to match that which was sent by the server.

There are various reasons for this, but from what I've seen, it mostly boils down to locales and invalid HTML.

With locales you can find hacks and tricks to hide the server rendered content for a moment, for example with CSS visibility, and then swap it with what your code calculates on the locale.

With invalid HTML it gets a bit trickier because the client side content might not even be structured as you expected. It potentially creates unreliable code. I've seen examples where some elements are rendered twice, though these are extreme cases. 

The efficient way is to go through these errors and fix them. 

You can use https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning to mark the spots where locales and other data that diverges, and fix them later!

For invalid HTML you ought to fix it, a tool like https://caninclude.glitch.me/ can help you make an argument as to why you ought to fix the semantics.

An option I am leaving out is deferring to the client, but you might want to consider that, though SEO, or rather, the HTML seen by bots/crawlers and non-JS entities might~will~ not see this content.

## Edit

A while back I opened an issue on the React repository regarding, why the server/client HTML reconciliation did not fail on the same way, rather than output divergent HTML, https://github.com/facebook/react/issues/24519, from what I got there, it is expected to be like this because of how browsers receive HTML.
